core/types/vote.go

```diff
@@ -0,0 +1,85 @@
+package types
+
+import (
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/crypto/bls"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	BLSPublicKeyLength = 48
+	BLSSignatureLength = 96
+)
+
+type BLSPublicKey [BLSPublicKeyLength]byte
+type BLSSignature [BLSSignatureLength]byte
+type ValidatorsBitSet uint64
+
+type VoteData struct {
+	SourceNumber uint64
+	SourceHash   common.Hash
+	TargetNumber uint64
+	TargetHash   common.Hash
+}
+
+func (d *VoteData) Hash() common.Hash { return rlpHash(d) }
+
+type VoteEnvelope struct {
+	VoteAddress BLSPublicKey
+	Signature   BLSSignature
+	Data        *VoteData
+
+	// caches
+	hash atomic.Value
+}
+
+type VoteAttestation struct {
+	VoteAddressSet ValidatorsBitSet
+	AggSignature   BLSSignature
+	Data           *VoteData
+	Extra          []byte
```

Comment:
what's `Extra` for, i cannot find the usage of it?

Reply:
Add comments to explain this